html/js/tree/Node.js

```diff
@@ -37,38 +38,105 @@ GTE.TREE = (function (parentModule) {
+                  .y(this.y)
+                  .click(function() {
+                      thisNode.onClick();
+                  });
+        if (this.player) {
+            this.shape.fill(this.player.colour);
+            this.drawPlayer();
+        } else {
+            this.shape.fill(GTE.COLOURS.BLACK);
+        }
+
+        if (GTE.MODE === GTE.MODES.PLAYERS && this.isLeaf()) {
+            this.shape.hide();
+        }
+    };
+
+    /**
+    * Draws the player. It needs to be done within the Node so that there is
+    * an instance of ContentEditable per Node
+    */
+    Node.prototype.drawPlayer = function () {
+        var thisPlayer = this.player;
+        this.playerNameText = thisPlayer.draw(
+            this.x + GTE.CONSTANTS.TEXT_NODE_MARGIN, this.y);
+        if (this.player.id === 0 && !GTE.tree.showChanceName) {
+            this.playerNameText.hide();
+        }
+    };
+
+    /**
+    * Toggles the visibility of the default name text
+    */
+    Node.prototype.togglePlayerName = function () {
+        if (this.playerNameText.visible() === false) {
+            this.playerNameText.show();
+        } else {
+            this.playerNameText.hide();
+        }
+    };
+
+    /**
+    * Updates player name. It sets the content editable text to the current
+    * player name
+    */
+    Node.prototype.updatePlayerName = function () {
+        this.playerNameText.setText(this.player.name);
     };
 
     /**
     * Function that defines the behaviour of the node on click
     */
     Node.prototype.onClick = function () {
-        if (GTE.MODE === GTE.MODES.ADD){
-            if (this.isLeaf()) {
-                // Always start with two nodes
+        switch (GTE.MODE) {
+            case GTE.MODES.ADD:
+                if (this.isLeaf()) {
+                    // Always start with two nodes
+                    GTE.tree.addChildNodeTo(this);
+                }
                 GTE.tree.addChildNodeTo(this);
-            }
-            GTE.tree.addChildNodeTo(this);
-        } else {
-            // If it is a leaf, delete itself, if not, delete all children
-            if (this.isLeaf()) {
-                this.delete();
-            } else {
-                GTE.tree.deleteChildrenOf(this);
-            }
+                // Tell the tree to redraw itself
+                GTE.tree.draw();
+                break;
+            case GTE.MODES.DELETE:
+                // If it is a leaf, delete itself, if not, delete all children
+                if (this.isLeaf()) {
+                    this.delete();
+                } else {
+                    GTE.tree.deleteChildrenOf(this);
+                }
+                GTE.tree.draw();
+                break;
+            case GTE.MODES.PLAYERS:
+                if (!this.isLeaf()) {
+                    // If player name is empty and default name is hidden,
+                    // show the default name
+                    if (this.player !== undefined) {
```

Comment:
Bug: This can throw an uncaught typeerror in line 126.

Steps to reproduce: 
1. Create a tree
2. add a new player (green)
3. Assign green player to one node
4. Remove green player
5. Try to assign Chance player to the node from step 3

To solve it, just check if this.player !== undefined && this.player !== null